<center> **Q3 **</center>
  <center>  **Data Analysis** </center>

**PART 1**

Lets import the data:

In [1]:
import pandas as pd
pd.options.mode.chained_assignment = None
customers = pd.read_csv('customer_data.csv')
requests = pd.read_csv('request_data.csv')


The file **customer_data.csv** contains the customers information, most importantly whether the customer_id was fraudulent or not and whether we suspected the customer to be fradulent or not.

The file **request_data.csv** contains the transfer requests the customers made.

Both the files combined contains information about normal and fraudulent customers, so we can explore the different features what might point to a fraudulent customer and potentially can create a simple model which can help us classify normal/fraudulent customers.

Lets join the customers and requests dataframe, for answering part 2.

**PART 2**

Lets check and remove the duplicates from customers and requests dataframes:

In [2]:
customers.duplicated().sum()

39

In [3]:
requests.duplicated().sum()

0

Since customers have duplicates, lets remove them:

In [4]:
customers = customers[~customers.duplicated()]

In [5]:
customers.duplicated().sum()

0

Lets join customers and requests for joint analysis:

In [6]:
joineddf = customers.merge(requests, on='customer_id', how='inner')

Now we can explore the fraudulent customers and their requests by filtering the **joineddf** to **frauddf** for fradulent customers.

As is clear from below summary of **frauddf**, we can make below infereneces:

1. On an average the fradulent customers managed to make around 2 transactions.
2. The average age of fraudlent customer is around 47 years.
3. On an average the cst profile was active for 20 days.
4. Average amount the fraudulent customer requested was 891 Euro's.

In [7]:
frauddf = joineddf[joineddf.fraudulent_cst==1]
frauddf.describe()

cst_successful_transfers_cnt     cst_age  cst_profile_age_days  \
count                    149.000000  149.000000            149.000000   
mean                       1.785235   46.946309             20.463087   
std                        1.216796   14.778447             25.606039   
min                        1.000000   19.000000              7.000000   
25%                        1.000000   34.000000             13.000000   
50%                        1.000000   46.000000             19.000000   
75%                        2.000000   59.000000             24.000000   
max                        7.000000   81.000000            321.000000   

       fraudulent_cst  suspicious_cst   amount_eur  
count           149.0      149.000000   149.000000  
mean              1.0        0.563758   890.722013  
std               0.0        0.497591   777.620694  
min               1.0        0.000000     5.000000  
25%               1.0        0.000000   276.310000  
50%               1.0        1.000000   670.140000  
75%               1.0        1.000000  1345.630000  
max               1.0        1.000000  4294.810000

Now lets compare these features for non-fraudulent customers:

In [8]:
nfrauddf = joineddf[joineddf.fraudulent_cst==0]
nfrauddf.describe()

cst_successful_transfers_cnt      cst_age  cst_profile_age_days  \
count                   2000.000000  1852.000000            2000.00000   
mean                      15.761000    35.626890             222.20350   
std                       76.142917    11.966394             215.39989   
min                        1.000000    18.000000               7.00000   
25%                        2.000000    27.000000              39.00000   
50%                        5.000000    32.000000             164.00000   
75%                       11.000000    41.000000             337.00000   
max                     1307.000000    85.000000            1284.00000   

       fraudulent_cst  suspicious_cst    amount_eur  
count          2000.0     2000.000000   2000.000000  
mean              0.0        0.049000   1691.055745  
std               0.0        0.215922   4811.542584  
min               0.0        0.000000      2.690000  
25%               0.0        0.000000    188.180000  
50%               0.0        0.000000    520.090000  
75%               0.0        0.000000   1387.495000  
max               0.0        1.000000  98682.290000

1. On an average the non-fradulent customers made around 16 transactions.
2. The average age of non-fraudlent customer is around 36 years.
3. On an average the cst profile was active for 222 days.
4. Average amount the fraudulent customer requested was 1691 Euro's.

**So in comparison, we can say that fradulent customers make less requests, are usually more older than non-fradulent customers, their profile is quite new, and they request almost half the amount than non-fradulent customers.**

Below are the summary statistics for the customer which our system did not suspect to be fraudulent. As we can observe they have almost the same feastures as the fraudulent customers on an average.

In [9]:
missed_customers = joineddf[(joineddf.fraudulent_cst == 1)&(joineddf.suspicious_cst == 0)]
missed_customers.describe()

cst_successful_transfers_cnt    cst_age  cst_profile_age_days  \
count                     65.000000  65.000000             65.000000   
mean                       1.830769  49.276923             24.261538   
std                        1.083531  15.749512             37.863768   
min                        1.000000  19.000000              7.000000   
25%                        1.000000  37.000000             15.000000   
50%                        2.000000  46.000000             21.000000   
75%                        2.000000  61.000000             25.000000   
max                        7.000000  81.000000            321.000000   

       fraudulent_cst  suspicious_cst   amount_eur  
count            65.0            65.0    65.000000  
mean              1.0             0.0  1038.822769  
std               0.0             0.0   762.787909  
min               1.0             0.0     5.000000  
25%               1.0             0.0   411.640000  
50%               1.0             0.0  1004.000000  
75%               1.0             0.0  1410.000000  
max               1.0             0.0  3983.240000

Below are the summary statistics for the customer which our system did suspect to be fraudulent. As we can observe they have almost the same feastures as the fraudulent customers on an average.

In [10]:
picked_customers = joineddf[(joineddf.fraudulent_cst == 1)&(joineddf.suspicious_cst == 1)]
picked_customers.describe()

cst_successful_transfers_cnt    cst_age  cst_profile_age_days  \
count                     84.000000  84.000000             84.000000   
mean                       1.750000  45.142857             17.523810   
std                        1.316027  13.806880              6.604993   
min                        1.000000  19.000000              7.000000   
25%                        1.000000  34.000000             11.000000   
50%                        1.000000  42.000000             17.000000   
75%                        2.000000  58.000000             24.000000   
max                        5.000000  74.000000             29.000000   

       fraudulent_cst  suspicious_cst   amount_eur  
count            84.0            84.0    84.000000  
mean              1.0             1.0   776.120238  
std               0.0             0.0   774.034025  
min               1.0             1.0    10.000000  
25%               1.0             1.0   201.840000  
50%               1.0             1.0   539.060000  
75%               1.0             1.0   963.470000  
max               1.0             1.0  4294.810000

Now lets explore some non-numeric features:

In [11]:
frauddf.columns

Index(['customer_id', 'cst_successful_transfers_cnt', 'cst_age',
       'customer_address_country', 'cst_profile_age_days', 'fraudulent_cst',
       'suspicious_cst', 'request_id', 'request_date', 'source_currency',
       'destination_currency', 'target_recipient_id', 'payment_type',
       'amount_eur'],
      dtype='object')

As we can see the most frequent coutries from which the fraudulent customers come are **gbr,deu, usa and irl**, which coincide with the fact that most customers in this dataset are from these countries.

In [12]:
frauddf.customer_address_country.value_counts()

gbr    66
deu    40
usa    15
irl     8
swe     5
ita     3
nor     2
phl     1
dnk     1
aus     1
egy     1
ltu     1
lux     1
bel     1
fra     1
esp     1
nld     1
Name: customer_address_country, dtype: int64

In [13]:
nfrauddf.customer_address_country.value_counts()

gbr    1159
usa     152
esp     114
deu      97
fra      58
irl      47
aus      36
nld      35
pol      28
ita      24
swe      23
hun      23
dnk      22
che      21
bel      20
est      14
bgr      13
fin      12
ind      12
nor      11
aut      11
can       9
prt       7
gib       6
cze       6
rou       6
svk       5
nzl       4
sgp       4
cyp       3
lux       3
mys       3
mlt       3
mex       1
srb       1
ggy       1
geo       1
grc       1
nga       1
hkg       1
ncl       1
are       1
Name: customer_address_country, dtype: int64

From below data we can see that the common source currency fraudulent customers request for transfer are **GBP, EUR and USD**.

In [14]:
frauddf.source_currency.value_counts()

GBP    65
EUR    62
USD    13
SEK     5
NOK     2
DKK     1
AUD     1
Name: source_currency, dtype: int64

In [15]:
nfrauddf.source_currency.value_counts()

GBP    1248
EUR     420
USD     150
AUD      41
SEK      30
CHF      23
DKK      23
PLN      18
HUF      16
NOK      11
CZK       9
BGN       7
RON       3
GEL       1
Name: source_currency, dtype: int64

We can see from below that the faviorite destination currencies for the fraudulent customers are: **PHP, INR EUR and GBP**. In comparison non-fraudulent customers commonly transfer to **EUR,GBP,INR,USD and HUF**. So we should look at the customers transferring to **PHP**.

In [16]:
frauddf.destination_currency.value_counts()

PHP    33
INR    29
EUR    28
GBP    21
NGN     8
BRL     5
IDR     4
USD     4
HUF     3
PLN     3
PKR     3
CZK     2
AUD     2
MAD     1
UAH     1
TRY     1
HKD     1
Name: destination_currency, dtype: int64

In [17]:
nfrauddf.destination_currency.value_counts()

EUR    716
GBP    290
INR    260
USD    162
HUF    126
PLN     87
AUD     52
BGN     34
CAD     34
BRL     32
RON     24
CZK     19
NZD     19
SEK     17
MYR     15
TRY     14
NGN     14
PKR     13
PHP     11
THB      9
SGD      9
HKD      8
DKK      8
CHF      6
MXN      4
CLP      4
NOK      4
UAH      2
JPY      2
AED      2
IDR      1
MAD      1
GEL      1
Name: destination_currency, dtype: int64

From below, we can infer that the common fraud payment type is **card**, which is also the common method for normal reuqests, so we should focus on card transactions for fraudlent behaviour.

In [18]:
frauddf.payment_type.value_counts()

card    120
bank     16
ach      13
Name: payment_type, dtype: int64

In [19]:
nfrauddf.payment_type.value_counts()

card    1100
bank     771
ach      129
Name: payment_type, dtype: int64

Now, lets create a simple logistic model to classify fraudulent and non-fradulent customers.

Lets import required modules and create a dataframe with suitable numeric features:

In [20]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report
from sklearn.model_selection import train_test_split

In [21]:
tdf  = joineddf[['customer_id','request_id', 'cst_age','cst_profile_age_days','suspicious_cst','payment_type','amount_eur','fraudulent_cst']]
tdf.dropna(inplace=True)
tdf = tdf.join(pd.get_dummies(tdf.payment_type))

In [22]:
X = tdf[['cst_age', 'cst_profile_age_days',
       'suspicious_cst', 'amount_eur', 'ach',
       'bank', 'card']]
y = tdf['fraudulent_cst']

Normalize numeric features:

In [23]:
for feature in ['cst_age','cst_profile_age_days','amount_eur']:
        X[feature] -= X[feature].mean()
        X[feature] /= X[feature].std()

In [24]:
model = LogisticRegression()
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

In [25]:
print(classification_report(y_test, y_pred))

             precision    recall  f1-score   support

          0       0.96      0.99      0.97       365
          1       0.83      0.53      0.64        36

avg / total       0.94      0.95      0.94       401



As we can see from above report, the model can precisely predict the fraudulent customer 83% of the times, which is not that bad.

Now let's see the features importance for this model and make some inference.

In [26]:
model.coef_

array([[ 0.85307096, -3.23340572,  2.58254703, -0.12428849, -1.66021923,
        -1.86843589, -0.67744013]])

In [27]:
X.columns

Index(['cst_age', 'cst_profile_age_days', 'suspicious_cst', 'amount_eur',
       'ach', 'bank', 'card'],
      dtype='object')

As we can see that the most important featrures are suspicious_cst and cst_age.

**PART 3**

**From all the anaylysis above and the logistic model, I can say that we should look for cst_age,suspicious_cst and if the destination currency is PHP if we wish to find the fraudulent customers and check their data before proceeding with the transaction, if these features fall in the fradulent zone we should stop that transaction and do manual checks to make sure its not fraudulent and if not then we should let the transaction happen automatically, to minimize the impact on our good customers. We can use the logistic model for automation too, but based on the amount of data, I would rather train this model again with more data or create more sophisticated model.**